File: packages/core/src/stream.ts
import {
  fetchNdJSON,
  type NdJSONFetchRequestInit,
  NdJSONStreamResponse,
  type NDJSONStreamResponseInit,
  PlainJsonError
} from './ndjson'
import { diffApply, diffCreate, type ObjectPatchDiff } from './diff'
import { type CompressedObjectStream, deflate, inflate } from './zip'

/**
 * A message in the object stream.
 */
export type ObjectStream<T extends object, E> =
  /** Initial full object state. */
  | { t: 'init'; d: T }
  /** Incremental diff operations. */
  | { t: 'delta'; d: ObjectPatchDiff }
  /** Stream completion signal with optional final state. */
  | { t: 'done'; d?: T }
  /** Custom application event. */
  | { t: 'event'; d: E }

/**
 * Wrapper for custom application events.
 */
export class StreamEvent<E> {
  constructor(readonly data: E) {}
}

/**
 * Wrapper to trigger a full re-initialization of the object.
 */
export class StreamReinit<T extends object> {
  constructor(readonly data: T) {}
}

/**
 * An async iterable of object states or stream control messages.
 */
export type ObjectStreamIterable<T extends object, E> = AsyncIterable<
  T | StreamReinit<T> | StreamEvent<E>
>

export interface ObjectStreamResponseInit extends NDJSONStreamResponseInit {
  /** Whether to send the final object state when the stream completes. */
  sendDataOnDone?: boolean
  /** Whether to use DCP compression. Defaults to true. */
  compressed?: boolean
}

/**
 * A Response implementation for the Differential Context Protocol (DCP).
 */
export class ObjectStreamResponse<T extends object, E = any> extends NdJSONStreamResponse<
  ObjectStream<T, E> | CompressedObjectStream<T, E>
> {
  /**
   * Creates a new ObjectStreamResponse.
   *
   * @param stream - The stream of objects and events to serve.
   * @param init - Optional configuration for the DCP stream.
   */
  constructor(stream: ObjectStreamIterable<T, E>, init?: ObjectStreamResponseInit) {
    async function* wrapped(): AsyncIterable<ObjectStream<T, E>> {
      let last: T | undefined = undefined
      for await (const chunk of stream) {
        // Send custom events
        if (chunk instanceof StreamEvent) {
          yield { t: 'event', d: chunk.data }
          continue
        }

        // Perform reinit
        if (chunk instanceof StreamReinit) last = undefined
        const data = chunk instanceof StreamReinit ? chunk.data : chunk

        if (last === undefined) {
          yield { t: 'init', d: data }
        } else {
          yield { t: 'delta', d: diffCreate<T>(last, data) }
        }

        last = JSON.parse(JSON.stringify(data))
      }

      // Finalize
      yield init?.sendDataOnDone ? { t: 'done', d: last } : { t: 'done' }
    }

    const compress = init?.compressed !== false
    const res = compress ? deflate(wrapped()) : wrapped()
    super(res, {
      ...init,
      headers: {
        ...init?.headers,
        'X-Diff-Version': compress ? '1; compressed' : '1'
      }
    })
  }
}

export interface ObjectStreamRequestInit<T extends object, E> extends NdJSONFetchRequestInit<
  ObjectStream<T, E> | CompressedObjectStream<T, E>
> {
  /** Called for each new data sample received. */
  onData?: (data: T) => void
  /** Called for each custom event received. */
  onEvent?: (event: E) => void
  /** Triggered when new data frames (init, delta, event, done) are received. */
  onFrame?: (frame: ObjectStream<T, E>) => void
  /** If the client accepts plain JSON responses as valid states. Defaults to true. */
  fallbackPlainJson?: boolean
}

/**
 * Fetches and processes a DCP object stream.
 *
 * @param input - The URL or Request to fetch.
 * @param init - Optional configuration for fetch and DCP processing.
 * @returns An async iterable that yields reconstructed object states.
 */
export async function* fetchObjectStream<T extends object, E = any>(
  input: string | URL | Request,
  init?: ObjectStreamRequestInit<T, E>
): AsyncIterable<T> {
  const { onEvent, ...ndInit } = init ?? {}
  let done: boolean = false
  let dataValue: T | undefined = undefined

  try {
    const res = fetchNdJSON<CompressedObjectStream<T, E>>(input, ndInit)
    for await (const data of inflate<T, E>(res)) {
      init?.onFrame?.(data)
      if (data.t === 'init') {
        dataValue = data.d
        init?.onData?.(dataValue)
        yield dataValue
      } else if (data.t === 'delta') {
        dataValue = diffApply<T>(dataValue, data.d)
        init?.onData?.(dataValue)
        yield dataValue
      } else if (data.t === 'event') {
        onEvent?.(data.d)
      } else if (data.t === 'done') {
        if (data.d !== undefined) dataValue = data.d
        done = true
      }
    }
  } catch (error) {
    // Treat `application/json` as the final data response. This makes it
    // easier to treat legacy APIs as DCP streaming ones
    if (error instanceof PlainJsonError && init?.fallbackPlainJson !== false) {
      return error.data
    } else {
      throw error
    }
  }

  // Finalize the stream.
  if (!dataValue) {
    throw Error(`Object stream ended without any data`)
  } else if (!done) {
    throw Error(`Object stream ended without done signal`)
  }
  return dataValue
}

File: python/diffcp/stream.py
from collections.abc import AsyncIterable, Callable
from typing import Any, TypeVar

from pydantic import BaseModel
from pydantic_core import to_jsonable_python

from .diff import diff_apply, diff_create
from .ndjson import NDJSONStreamingResponse, PlainJsonError, fetch_ndjson
from .zip import deflate, inflate

T = TypeVar('T')
E = TypeVar('E')


class StreamEvent:
    """
    Emit an event via DCP.
    """

    def __init__(self, data: Any):
        self.data = data


class StreamReinit:
    """
    Force a complete object update.
    """

    def __init__(self, data: Any):
        self.data = data


async def fetch_object_stream[T, E](
    url: str,
    *,
    on_data: Callable[[T], None] | None = None,
    on_event: Callable[[E], None] | None = None,
    on_frame: Callable[[dict[str, Any]], None] | None = None,
    fallback_plain_json: bool = True,
    **kwargs: Any,
) -> AsyncIterable[T]:
    data_value: Any = None
    done = False

    try:
        nd_stream = fetch_ndjson(url, **kwargs)
        async for frame in inflate(nd_stream):
            if on_frame:
                on_frame(frame)

            t = frame['t']
            d: Any = frame.get('d')

            if t == 'init':
                data_value = d
                if on_data:
                    on_data(data_value)
                yield data_value
            elif t == 'delta':
                data_value = diff_apply(data_value, d)
                if on_data:
                    on_data(data_value)
                yield data_value
            elif t == 'event':
                if on_event:
                    on_event(d)
            elif t == 'done':
                if d is not None:
                    data_value = d
                done = True
    except PlainJsonError as e:
        if fallback_plain_json:
            yield e.data
            return
        else:
            raise e

    if data_value is None:
        raise Exception('Object stream ended without any data')
    if not done:
        raise Exception('Object stream ended without done signal')


class ObjectStreamResponse(NDJSONStreamingResponse):
    """
    Differential Context Protocol (DCP) streaming response.
    """

    def __init__(
        self,
        content: AsyncIterable[T | StreamReinit | StreamEvent],
        status_code: int = 200,
        headers: dict[str, str] | None = None,
        ping_interval: int | float = 15,
        send_data_on_done: bool = False,
        compressed: bool = False,
        **kwargs: Any,
    ) -> None:
        async def _diff_wrapper() -> AsyncIterable[dict[str, Any]]:
            last: Any | None = None
            async for chunk in content:
                # Send custom events
                if isinstance(chunk, StreamEvent):
                    event_data = chunk.data
                    if isinstance(event_data, BaseModel):
                        event_data = event_data.model_dump(mode='json', by_alias=True)
                    yield {'t': 'event', 'd': to_jsonable_python(event_data)}
                    continue

                # Perform reinit
                if isinstance(chunk, StreamReinit):
                    chunk = chunk.data
                    last = None

                # Handle Pydantic models
                if isinstance(chunk, BaseModel):
                    current = chunk.model_dump(mode='json', by_alias=True)
                else:
                    current = to_jsonable_python(chunk)

                if last is None:
                    yield {'t': 'init', 'd': current}
                else:
                    diff = diff_create(last, current)
                    if len(diff) > 0:
                        yield {'t': 'delta', 'd': diff}
                last = current

            # Finalize
            if send_data_on_done and last is not None:
                yield {'t': 'done', 'd': last}
            else:
                yield {'t': 'done'}

        # Apply compression if requested
        wrapped_content = deflate(_diff_wrapper()) if compressed else _diff_wrapper()

        super().__init__(
            content=wrapped_content,
            status_code=status_code,
            headers={'X-Diff-Version': '1; compressed' if compressed else '1', **(headers or {})},
            ping_interval=ping_interval,
            **kwargs,
        )
